turbopack/crates/turbo-tasks/src/value_type.rs

```diff
@@ -182,16 +181,46 @@ impl ValueType {
     }
 
     /// This is internally used by `#[turbo_tasks::value_impl]`
-    pub fn register_trait(&mut self, trait_type: TraitTypeId) {
-        self.traits.insert(trait_type);
+    pub fn register_trait<T: ?Sized>(
+        &mut self,
+        trait_type: TraitTypeId,
+        meta: std::ptr::DynMetadata<T>,
+    ) where
+        Box<T>: VcValueTrait,
+    {
+        let meta = unsafe {
+            // Safety: DynMetadata does not store T. (And it can't, because T of DynMetadata is
+            // not Sized)
+            std::mem::transmute::<DynMetadata<T>, DynMetadata<dyn Any + Send + Sync>>(meta)
+        };
+
+        self.traits.insert(trait_type, meta);
+    }
+
+    /// Composes a fat pointer from a `dyn Any` for the given trait_id
+    pub(crate) fn as_trait_ptr<T>(
+        &self,
+        trait_type: TraitTypeId,
+        ptr: *const (dyn Any + Send + Sync),
+    ) -> *const T
+    where
+        T: std::ptr::Pointee<Metadata = std::ptr::DynMetadata<T>> + ?Sized,
+    {
```

Comment:
As-is, this method needs to unsafe because there's no guarantee that the caller's `TraitTypeId` and `T` match up. Using `Box<dyn Any>` to store the `DynMetadata` would let you do a checked cast.

Reply:
i can solve this by just moving the call to `get_trait_type_id` into this function, so that the relationship between the two is clear.  And i can modify register trait in the same way.

I believe that solves the safety issue